pfcp/adapter/adapter.go

```diff
@@ -257,8 +238,64 @@ func HandlePfcpSessionEstablishmentResponse(msg *udp.Message) {
 	}
 
 	// Get N3 interface UPF
+	defaultPath := smContext.Tunnel.DataPathPool.GetDefaultPath()
+	if defaultPath == nil {
+		logger.PfcpLog.Errorln("failed to get default path")
+		return
+	}
 	ANUPF := smContext.Tunnel.DataPathPool.GetDefaultPath().FirstDPNode
 
+	if rsp.CreatedPDR != nil {
+		ueIPAddress := FindUEIPAddress(rsp.CreatedPDR)
+		if ueIPAddress != nil {
+			smContext.SubPfcpLog.Infof("upf provided ue ip address [%v]", ueIPAddress)
+			// Release previous locally allocated UE IP-Addr
+			err := smContext.ReleaseUeIpAddr()
+			if err != nil {
+				logger.PfcpLog.Errorf("failed to release UE IP-Addr: %+v", err)
+			}
+
+			// Update with one received from UPF
+			smContext.PDUAddress.Ip = ueIPAddress
+			smContext.PDUAddress.UpfProvided = true
+		}
+
+		// Store F-TEID created by UPF
+		fteid, err := FindFTEID(rsp.CreatedPDR)
+		if err != nil {
+			logger.PfcpLog.Errorf("failed to parse TEID IE: %+v", err)
+			return
+		}
+		logger.PfcpLog.Infof("created PDR FTEID: %+v", fteid)
+		ANUPF.UpLinkTunnel.TEID = fteid.TEID
+		upf := context.RetrieveUPFNodeByNodeID(*nodeID)
+		if upf == nil {
+			logger.PfcpLog.Errorf("can't find UPF[%s]", nodeID.ResolveNodeIdToIp().String())
+			return
+		}
+		upf.N3Interfaces = make([]context.UPFInterfaceInfo, 0)
+		n3Interface := context.UPFInterfaceInfo{}
+		n3Interface.IPv4EndPointAddresses = append(n3Interface.IPv4EndPointAddresses, fteid.IPv4Address)
+		upf.N3Interfaces = append(upf.N3Interfaces, n3Interface)
+	}
+	smContext.SMLock.Unlock()
```

Comment:
Could you pls help with corresponding Lock statement? Also its preferred to use defer statement to avoid unintentional stale lock.

Reply:
This was missing, I just added it.

Reply:
there are multiple return statements in between. Could you please do following

mutex.lock()
defer mutex.unlock()

please see - https://gobyexample.com/mutexes

Reply:
Good catch, done